source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc

```diff
@@ -109,13 +109,15 @@ void GradientController::enterMinRTTSamplingWindow() {
 }
 
 void GradientController::updateMinRTT() {
-  ASSERT(inMinRTTSamplingWindow());
```

Comment:
You're removing the assertion because the race condition triggers it?

Reply:
yes, the assertion caused the crash in the unit test.

Reply:
This is asserting an invariant. If the unit test is triggering this, then it is doing something incorrectly. You were not seeing assertion failures in production, right? This needs to remain.

Reply:
But i am not sure whether it should be invariant, let's simplify the logic in this way:
```
if(inMinRTTSamplingWindow() && ...) {
	// start invoking updateMinRTT()
	ASSERT(inMinRTTSamplingWindow());
	...
	deferred_limit_value_.store(0);  // set it not in the minRTT sampling windows.
	...
}
```
If multiple threads execute the above code code snippet simultaneously, i think the assertion might fail.

Regarding why we didn't see the assertion failure in production, i'm guessing that's related to the worker threads in prod are pending on the [sample_mutation_mtx_ mutex lock](https://github.com/envoyproxy/envoy/blob/114d5aea7b05ef697d4fcd94201fe090a19077b7/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L115), which executes after the assertion. But in the unit test, they pends on [syncpoint](https://github.com/envoyproxy/envoy/blob/07abab6a106328fa90ce0659723bf4185d92cc4c/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L227), which executes before the assertion. When we signal the syncpoint in the unit test, it will cause assertion failure as one thread has already set it not in the minRTT sampling window previously.